controllers/node_controller.go

```diff
@@ -0,0 +1,177 @@
+			k8sclientset:       clientset,
+			clusterServiceCIDR: clusterConfig.Network().GetServiceCIDR(),
+			watchNamespace:     watchNamespace,
+			recorder:           mgr.GetEventRecorderFor(NodeController),
+		},
+	}, nil
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which reads that state of the cluster for a
+// CertificateSigningRequests object and aims to move the current state of the cluster closer to the desired state.
+func (r *nodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	_ = r.log.WithValues(NodeController, req.NamespacedName)
+	// Prevent WMCO upgrades while Node objects are being processed
+	if err := condition.MarkAsBusy(r.client, r.watchNamespace, r.recorder, NodeController); err != nil {
+		return ctrl.Result{}, err
+	}
+	defer func() {
+		err = markAsFreeOnSuccess(r.client, r.watchNamespace, r.recorder, NodeController, result.Requeue, err)
+	}()
+
+	// Fetch Node reference
+	node := &core.Node{}
+	if err := r.client.Get(ctx, req.NamespacedName, node); err != nil {
+		if k8sapierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - return error to requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Create a new signer using the private key that the instances will be configured with
+	r.signer, err = signer.Create(types.NamespacedName{Namespace: r.watchNamespace,
+		Name: secrets.PrivateKeySecret}, r.client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to create signer from private key secret: %w", err)
+	}
+
+	if _, ok := node.GetAnnotations()[metadata.RebootAnnotation]; ok {
+		r.log.Info("rebooting node", "name", node.Name)
+		if err := r.safeReboot(ctx, node); err != nil {
+			return ctrl.Result{}, fmt.Errorf("full instance reboot failed for node %s: %w", node.Name, err)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// safeReboot safely restarts the instance associated with the given node object, first cordoning and draining the node
+func (r *nodeReconciler) safeReboot(ctx context.Context, node *core.Node) error {
+	instanceInfo, err := r.instanceFromNode(node)
+	if err != nil {
+		return err
+	}
+	nc, err := nodeconfig.NewNodeConfig(r.client, r.k8sclientset, r.clusterServiceCIDR, r.watchNamespace,
+		instanceInfo, r.signer, nil, nil, r.platform)
+	if err != nil {
+		return fmt.Errorf("failed to create new nodeconfig: %w", err)
+	}
+
+	drainer := nc.NewDrainHelper()
+	if err := drain.RunCordonOrUncordon(drainer, node, true); err != nil {
+		return fmt.Errorf("unable to cordon node %s: %w", node.Name, err)
+	}
+	if err := drain.RunNodeDrain(drainer, node.Name); err != nil {
+		return fmt.Errorf("unable to drain node %s: %w", node.Name, err)
+	}
```

Comment:
Instead of exporting NewDrainHelper, what do you think about keeping the drain logic contained within the Nodeconfig package, and moving `SafeReboot` to be a method of Nodeconfig?